bot/internal/bot/changelog.go

```diff
@@ -0,0 +1,123 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bot
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/slices"
+)
+
+const NoChangelogLabel string = "no-changelog"
+const ChangelogPrefix string = "changelog: "
+const ChangelogRegex string = "(?mi)^changelog: .*"
+
+// Checks if the PR contains a changelog entry in the PR body, or a "no-changelog" label.
+//
+// A few tests are performed on the extracted changelog entry to ensure it conforms to a
+// common standard.
+func (b *Bot) CheckChangelog(ctx context.Context) error {
+	pull, err := b.c.GitHub.GetPullRequest(ctx,
+		b.c.Environment.Organization,
+		b.c.Environment.Repository,
+		b.c.Environment.Number,
+	)
+	if err != nil {
+		return trace.Wrap(err, "failed to retrieve pull request for https://github.com/%s/%s/pull/%d", b.c.Environment.Organization, b.c.Environment.Repository, b.c.Environment.Number)
+	}
+
+	if slices.Contains(pull.UnsafeLabels, NoChangelogLabel) {
+		log.Printf("PR contains %q label, skipping changelog check", NoChangelogLabel)
+		return nil
+	}
+
+	changelogEntry, err := b.getChangelogEntry(ctx, pull.UnsafeBody)
+	if err != nil {
+		return trace.Wrap(err, "failed to get changelog entry")
+	}
+
+	err = b.validateChangelogEntry(ctx, changelogEntry)
+	if err != nil {
+		return trace.Wrap(err, "failed to validate changelog entry")
+	}
+
+	return nil
+}
+
+func (b *Bot) getChangelogEntry(ctx context.Context, prBody string) (string, error) {
+	changelogRegex := regexp.MustCompile(ChangelogRegex)
+
+	changelogMatches := changelogRegex.FindAllString(prBody, 2) // Two or more changelog entries is invalid, so no need to search for more than two.
+	if len(changelogMatches) == 0 {
+		return "", b.logFailedCheck(ctx, "Changelog entry not found in the PR body. Please add a %q label to the PR, or a changelog line starting with `%s` followed by the changelog entry for the PR.", NoChangelogLabel, ChangelogPrefix)
+	}
+
+	if len(changelogMatches) > 1 {
+		return "", b.logFailedCheck(ctx, "Found two or more changelog entries in the PR body: %v", changelogMatches)
+	}
+
+	changelogEntry := changelogMatches[0][len(ChangelogPrefix):] // Case insensitive prefix removal
+	log.Printf("Found changelog entry %q", changelogEntry)
+	return changelogEntry, nil
+}
+
+// Checks for common issues with the changelog entry.
+// This is not intended to be comprehensive, rather, it is intended to cover the majority of problems.
+func (b *Bot) validateChangelogEntry(ctx context.Context, changelogEntry string) error {
+	if strings.TrimSpace(changelogEntry) == "" {
+		return b.logFailedCheck(ctx, "The changelog entry must contain one or more non-whitespace characters")
```

Comment:
To make logic more clear, I would make the following changes:

- Instead of returning `b.logFailedCheck()` from each branch here and in "getChangelogEntry", have them return a normal `trace.BadParameter("xxx")` error.
- In `CheckChangelog` call `b.logFailedCheck(ctx, err)` when the respective function returns an error.

Basically, move `logFailedCheck` call up the stack. Otherwise it's very implicit that get/validate functions also leave comments on the PR.

Reply:
I disagree here, on the grounds that `validateChangelogEntry` could potentially return an error that is not due to an invalid changelog. Currently this is not an issue, however, if a developer were to add another check that does something like checking to see if the changelog entry is already in the CHANGELOG.md file, then this could result in an error message being written to the PR body that is not due to an invalid changelog entry. More specifically, if a check was added that did something like `err := os.ReadFile("CHANGELOG.md")` then this error could potentially get logged as a changelog entry failure.

Maybe one of these would be a better solution to the issue:
* I could add a comment to the `validateChangelogEntry` function that explicitly says that it logs errors to the PR
* I could return an error type that is specific to a changelog entry problem, and only log to the PR if the error is of this type

What do you think?

If I were to go with the second route, would it be better to return and instance of this new error type, or return a `trace.Wrap` of this new error type? In other words, `ValidationError("some message")` or `trace.Wrap(ValidationError("some message"), "some error occured")`?